### PR TITLE
feat(encrypt): selectable encryption engine + AES-GCM re-encryption migrator

### PR DESCRIPTION
**Pull request by rusackas**
> Phase 1 (engine selection) + Phase 2 (re-encryption migrator) of the proposal in [`docs/sip/authenticated-encryption-at-rest.md`](docs/sip/authenticated-encryption-at-rest.md). Both are **opt-in**; the default engine is unchanged, so existing installs are unaffected. Flipping the default for fresh installs (Phase 3) remains future work.

### SUMMARY

App-encrypted fields (database passwords, SSH tunnel credentials, OAuth tokens, …) use `sqlalchemy_utils.EncryptedType`, which defaults to **`AesEngine` (AES-CBC)** — unauthenticated encryption. **`AesGcmEngine` (AES-GCM)** is authenticated and fails closed on tampering. This PR makes the engine selectable **and** provides a migration path so existing deployments can adopt it safely.

**Phase 1 — engine selection**
- Adds `SQLALCHEMY_ENCRYPTED_FIELD_ENGINE` config (`"aes"` | `"aes-gcm"`), **defaulting to `"aes"`** — zero behavior change for existing installs.
- The default `SQLAlchemyUtilsAdapter` honors it; an explicit `engine` kwarg still wins. Unknown values fall back to AES-CBC.

**Phase 2 — CBC→GCM re-encryption migrator (new in this revision)**
- Teaches the existing `SecretsMigrator` (previously only used for `SECRET_KEY` rotation) an **engine-migration mode**: decrypt each value with the source engine under the current `SECRET_KEY`, re-encrypt with the target engine.
- Exposed via a new `--engine` flag on the existing CLI: `superset re-encrypt-secrets --engine aes-gcm`.
- Transactional (all-or-nothing) and **idempotent per column** — already-migrated values are skipped, so a run can be safely repeated or resumed. Engine-symmetric (supports `--engine aes` rollback) and composable with a concurrent key rotation (`--previous_secret_key`).
- `UPDATING.md` documents the operator runbook (backup → `re-encrypt-secrets --engine aes-gcm` → set config → restart).

The instance-wide default flip (Phase 3) is intentionally **not** included — it's the subject of the SIP.

### TESTING INSTRUCTIONS

```
pytest tests/unit_tests/utils/encrypt_test.py
```

Unit tests cover engine selection (default/explicit/fallback) **and** the migrator: CBC→GCM re-encryption, idempotency/resumability (already-GCM values skipped), reading CBC after the config is flipped first, and failure counting for undecryptable values. Existing `SecretsMigrator` key-rotation integration tests are preserved unchanged (the new `previous_secret_key`/`target_engine` signature is backward-compatible).

### ADDITIONAL INFORMATION

- [ ] Has associated issue:
- [ ] Required feature flags:
- [ ] Changes UI
- [ ] Includes DB Migration (no schema change; ciphertext is re-encrypted in place via the opt-in CLI)
- [x] Introduces new feature or API (opt-in config + CLI flag; no default change)
- [ ] Removes existing feature or API

🤖 Generated with [Claude Code](https://claude.com/claude-code)

### CODE REVIEW

File: superset/utils/encrypt.py [hunk 1]
```diff
@@ -177,6 +226,42 @@ def _select_columns_from_table(
...
+           current one in engine-migration mode) reading AES-CBC source data
+           after the config has already been flipped to the target engine.
+        """
+        decryptors = [encrypted_type]
+        if self._previous_secret_key:
+            decryptors.append(
+                EncryptedType(
+                    type_in=encrypted_type.underlying_type,
+                    key=self._previous_secret_key,
+                )
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/utils/encrypt.py]: Good catch, this is a real bug for AES-GCM deployments. Fixed in fc28e60: `_source_decryptors` now tries the previous key under the column's own engine (so GCM ciphertext from the old key decrypts on rotation) as well as the historical AES-CBC engine (which preserves the engine-flipped-first migration path). Added a unit test, `test_key_rotation_for_aes_gcm_column`, that fails without the fix.

File: superset/utils/encrypt.py [hunk 2]
```diff
@@ -177,6 +226,42 @@ def _select_columns_from_table(
...
+            )
+        return decryptors
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/utils/encrypt.py:360]: Same root cause as the other thread, and you're right. Fixed in fc28e60: the previous-key fallback now uses the column's configured engine (plus AES-CBC as a second attempt), so AES-GCM key rotation works. Covered by the new `test_key_rotation_for_aes_gcm_column` unit test.

File: superset/cli/update.py [hunk 3]
```diff
@@ -110,17 +110,37 @@ def update_api_docs() -> None:
     help="An optional previous secret key, if PREVIOUS_SECRET_KEY "
     "is not set on the config",
 )
-def re_encrypt_secrets(previous_secret_key: Optional[str] = None) -> None:
+@click.option(
+    "--engine",
+    "-e",
+    "target_engine_name",
+    required=False,
+    type=click.Choice(sorted(ENCRYPTION_ENGINES)),
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/cli/update.py]: Agreed, applied in fc28e60 — `click.Choice(..., case_sensitive=False)`. The engine keys are lowercase so the lookup still resolves correctly after Click normalizes the input.

File: superset/utils/encrypt.py [hunk 4]
```diff
@@ -65,6 +78,12 @@ def create(
         **kwargs: Optional[dict[str, Any]],
     ) -> TypeDecorator:
         if app_config:
+            # Select the encryption engine from config, defaulting to the
+            # historical AES-CBC engine for backward compatibility. An explicit
+            # ``engine`` kwarg (e.g. from the migrator) always takes precedence.
+            if "engine" not in kwargs:
+                engine_name = app_config.get("SQLALCHEMY_ENCRYPTED_FIELD_ENGINE", "aes")
+                kwargs["engine"] = ENCRYPTION_ENGINES.get(engine_name, AesEngine)
             return EncryptedType(*args, lambda: app_config["SECRET_KEY"], **kwargs)
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/utils/encrypt.py:157]: On the warning: good idea, added in fc28e60 — the adapter now logs a warning naming the valid engines before falling back to AES-CBC when `SQLALCHEMY_ENCRYPTED_FIELD_ENGINE` is unrecognized. On the mypy concern: `pre-commit run mypy` passes as-is on this assignment, so no change was needed there.

File: superset/config.py [hunk 5]
```diff
@@ -295,6 +295,16 @@ def _try_json_readsha(filepath: str, length: int) -> str | None:
...
 
+# Encryption engine used by the default SQLAlchemyUtilsAdapter for app-encrypted
+# fields. Options:
+#   "aes"     - AES-CBC (historical default; unauthenticated, queryable)
+#   "aes-gcm" - AES-GCM (authenticated encryption; recommended for NEW installs)
+# WARNING: changing this on a database that already holds encrypted secrets
+# (database passwords, SSH tunnel credentials, OAuth tokens, ...) will make
+# those values undecryptable unless they are re-encrypted first. See the
+# authenticated-encryption SIP/migration before switching an existing install.
+SQLALCHEMY_ENCRYPTED_FIELD_ENGINE: Literal["aes", "aes-gcm"] = "aes"
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/config.py:309]: Fair point, fixed in fc28e60 — the comment block above `SQLALCHEMY_ENCRYPTED_FIELD_TYPE_ADAPTER` now points operators at the `SQLALCHEMY_ENCRYPTED_FIELD_ENGINE` knob as the preferred way to switch engines, keeping the custom-adapter example as the advanced option.

File: superset/cli/update.py [hunk 6]
```diff
@@ -110,17 +110,37 @@ def update_api_docs() -> None:
...
-            "No previous secret key provided; nothing to re-encrypt.",
+            "No previous secret key or target engine provided; nothing to re-encrypt.",
             fg="yellow",
         )
         return
-    secrets_migrator = SecretsMigrator(previous_secret_key=previous_secret_key)
+    secrets_migrator = SecretsMigrator(
+        previous_secret_key=previous_secret_key,
+        target_engine=target_engine,
+    )
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/cli/update.py:151]: I don't think this one holds as stated. The repro requires `SQLALCHEMY_ENCRYPTED_FIELD_ENGINE` to be flipped to GCM *and* a stale `PREVIOUS_SECRET_KEY` left in config while the data is still CBC under the current key — but the documented runbook (and the `--engine` help text) says to run the migrator *before* flipping the config. More importantly, after the `_source_decryptors` fix in fc28e60 the previous-key fallback now also tries AES-CBC, so even the config-flipped-first case decrypts CBC source data under the previous key. Forwarding `PREVIOUS_SECRET_KEY` during an engine migration is intentional for operators doing a combined key-rotation + engine change. Leaving the CLI as-is.

File: superset/utils/encrypt.py [hunk 7]
```diff
@@ -65,6 +78,19 @@
...
+            # Select the encryption engine from config, defaulting to the
+            # historical AES-CBC engine for backward compatibility. An explicit
+            # ``engine`` kwarg (e.g. from the migrator) always takes precedence.
+            if "engine" not in kwargs:
+                engine_name = app_config.get("SQLALCHEMY_ENCRYPTED_FIELD_ENGINE", "aes")
+                if engine_name not in ENCRYPTION_ENGINES:
+                    logger.warning(
+                        "Unrecognized SQLALCHEMY_ENCRYPTED_FIELD_ENGINE %r;"
+                        " falling back to AES-CBC. Valid engines: %s",
+                        engine_name,
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/utils/encrypt.py]: Fixed in 731a476 — the unrecognized-engine warning no longer interpolates the config-sourced engine value into the log message; it just lists the valid engine names, which is enough to diagnose a typo. That removes the tainted-config-value-into-log flow CodeQL traced (the config dict also holds SECRET_KEY).

File: superset/cli/update.py [hunk 8]
```diff
@@ -110,17 +110,37 @@ def update_api_docs() -> None:
...
     try:
         stats = secrets_migrator.run()
     except Exception as exc:  # pylint: disable=broad-except
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/cli/update.py:154]: Good call — added in 0fe86ea. `tests/integration_tests/cli_tests.py` now covers the `--engine` CLI path with three tests: that `--engine aes-gcm` resolves to `AesGcmEngine` and is passed through to `SecretsMigrator` (with `previous_secret_key=None`), that the option is case-insensitive (`AES-GCM` resolves the same), and that an unrecognized engine name yields a Click usage error listing the valid choices. I patched the whole `SecretsMigrator` rather than just its `run` method so the assertion can inspect the constructor kwargs without needing real DB introspection.

**Review by richardfogaca — COMMENTED** [review]
_Posting on Richard's behalf — this is his PR reviewer agent. Forward any pushback to him and he'll loop me back in._

Really solid PR overall — opt-in end to end, genuinely zero-behavior-change by default, and reusing `SecretsMigrator` instead of growing a parallel tool is the right call. The four items under "Design" are the ones worth a look before merge (mostly fail-open vs fail-closed posture and the operational window); the rest are optional. All line numbers verified against HEAD `6671c90`.

### Design — worth discussing before merge

- **`superset/utils/encrypt.py:86-97`**

  An unrecognized `SQLALCHEMY_ENCRYPTED_FIELD_ENGINE` logs a warning and silently falls back to AES-CBC. For a crypto-strength selector that feels like the wrong default posture: an operator who typos `"aes_gcm"` or `"AES-GCM"` believes they have authenticated encryption but runs unauthenticated CBC indefinitely — and if the typo happens *after* a GCM migration, new secrets get written CBC into a GCM database while existing ones become unreadable. It's also inconsistent with the CLI, which rejects unknown names via `click.Choice` (case-insensitively) at `superset/cli/update.py:118`.

  Would it be worth raising at field-creation time instead of warning, so a misconfig fails at startup rather than degrading silently? (Or at minimum case-normalizing to match the CLI.)

- **`superset/utils/encrypt.py:343-349`**

  The "already in target form" fast path treats *any* successful decrypt under the target engine as proof a value is already migrated. Forward (CBC→GCM) that's sound — the GCM tag is a real authenticator. In the rollback direction (`--engine aes`), CBC decryption is unauthenticated, so a GCM blob can spuriously "decrypt" (length check + padding + UTF-8 all passing by accident) and get counted as `skipped` — left as GCM ciphertext that bricks once config is flipped back to `"aes"`, with the run still reporting success. Low probability per value, but it's nonzero across a large fleet, and the docstring on `test_engine_migration_gcm_to_cbc_rolls_back` (`tests/unit_tests/utils/encrypt_test.py:164`) suggests you were already thinking about this.

  Could the skip-check prefer the authenticated interpretation — e.g. when the target is CBC, only treat a value as "already migrated" if the GCM source decrypt *fails* first? Alternatively documenting rollback as best-effort would also close it.

- **`UPDATING.md:73-83`**

  Two gaps in the runbook window between step 2 (migrate) and step 4 (restart): (a) a live app keeps writing *new* secrets as CBC during that window, and those become undecryptable after the flip — the code already handles this fine (re-running the migrator after the restart sweeps them up, per `test_engine_migration_reads_cbc_after_config_already_flipped`), but the runbook never says to; (b) in multi-worker deployments there's an unavoidable decrypt-outage window between the migration commit and the last worker restarting with the new config, since runtime reads use only the single configured engine — "transactional" may read as "zero-downtime" to operators when it isn't.

  Small suggestion: add a final "re-run `re-encrypt-secrets --engine aes-gcm` after restart" step and a sentence about scheduling the cutover in a quiet window. WDYT?

- **`superset/utils/encrypt.py:204`**

  `discover_encrypted_fields` only matches `isinstance(col.type, EncryptedType)`, so a deployment using a custom `SQLALCHEMY_ENCRYPTED_FIELD_TYPE_ADAPTER` gets a silent no-op migration — 0 columns discovered, exit 0, "success". Pre-existing for key rotation, but this PR makes the migrator the blessed engine-migration path.

  Would it be worth detecting a non-default adapter and warning (or refusing)? Happy to keep as-is if that's considered out of scope.

---

### Other suggestions

- **`superset/utils/encrypt.py:85`**

  `app_config.get("SQLALCHEMY_ENCRYPTED_FIELD_ENGINE", "aes")` — since the key now has a default in `superset/config.py:309`, the house pattern is `app_config["..."]`; the default currently lives in three places (config.py, this `.get`, and the `ENCRYPTION_ENGINES.get(..., AesEngine)` fallback), which could drift when Phase 3 flips it. Totally optional given custom configs may omit the key.

- **`superset/cli/update.py:114`**

  The PR body advertises composability with key rotation (`--previous_secret_key` + `--engine` in one run), but no test exercises the combined mode (old-key CBC → current-key GCM). Could we add one? It's the mode most likely to regress since each mode's tests pin the other's variable.

- Few small nits, take or leave: function-local `AesGcmEngine` imports in `tests/integration_tests/cli_tests.py:375` (and again ~402) could move to module top; `engines: list[type[Any]]` at `superset/utils/encrypt.py:274` could be `type[EncryptionDecryptionBaseEngine]` since it's already imported from the same module.

### Praise

The unit-test suite is unusually thoughtful for crypto code — each test docstring states the operational scenario it protects, and `test_engine_migration_reads_cbc_after_config_already_flipped` (`tests/unit_tests/utils/encrypt_test.py:140`) covers the exact "operator did the steps in the wrong order" recovery path most migrators forget.

**Comment by rusackas** [comment]
Thanks @richardfogaca — really sharp review. Addressed the design items in 10c49b6 under one posture: **authenticated checks gate decisions; unauthenticated paths never silently win.**

**Unknown engine → fail-closed** (`encrypt.py:86-97`). Agreed this was the wrong default for a crypto selector. An unrecognized `SQLALCHEMY_ENCRYPTED_FIELD_ENGINE` now raises at field construction instead of degrading to CBC, and the value is case/separator-normalized (`aes_gcm`/`AES-GCM` → `aes-gcm`) to match the CLI's `click.Choice`. Kept the offending value out of the error message (same clear-text-logging rationale as the CodeQL fix in 731a476) but list the valid engines. Absent key still defaults to `aes` so existing installs are untouched. Mirrors house precedent (refusing to boot on a default SECRET_KEY) and kills the post-migration split-brain you flagged.

**Rollback spurious-skip → fixed, not documented** (`encrypt.py:343-349`). You're right that an unauthenticated CBC "success" shouldn't classify a value as migrated. When the target engine is unauthenticated, the migrator now probes the value under any authenticated engine (current + previous key) *first*; an authenticated decrypt wins and the value is re-encrypted rather than skipped. Generalized on *is the target authenticated* rather than hardcoding engine names. Since the coincidental CBC-decrypt-of-GCM can't be crafted deterministically, the new test pins the **ordering invariant** (forces the target read to succeed, asserts re-encrypt not skip). Explicitly declined the "document as best-effort" alternative — a run reporting success while leaving rows that brick on the next flip violates the migrator's own transactional/idempotent contract.

**Runbook** (`UPDATING.md`): added the post-restart re-run step (sweeps secrets written CBC during the migration window) and a sentence that multi-worker cutover has a brief decrypt-outage and isn't zero-downtime.

**Smaller items**: combined key-rotation + engine-migration is now covered at both the migrator (`encrypt_test.py`) and CLI (`cli_tests.py`) levels; de-duped the redundant previous-key decryptor when `prev_key == SECRET_KEY`; tightened the engine-list type hint; hoisted the function-local `AesGcmEngine` imports.

Two I left as follow-ups rather than widen this PR, lmk if you'd rather I fold them in:
- **`discover_encrypted_fields` custom-adapter no-op** (`encrypt.py:204`): a non-default `SQLALCHEMY_ENCRYPTED_FIELD_TYPE_ADAPTER` yields a silent 0-column migration. Pre-existing, but worth a warn/refuse — happy to file an issue.
- The `app_config[...]` vs `.get(...,"aes")` default: kept `.get` so configs that omit the key still work, but the silent `ENCRYPTION_ENGINES.get(..., AesEngine)` fallback is gone, so the default now lives in just config.py + the documented `DEFAULT_ENCRYPTION_ENGINE_NAME`.

**Review by richardfogaca — COMMENTED** [review]
_Posting on Richard's behalf - this is his PR reviewer agent. Forward any pushback to him and he'll loop me back in._

Left one functional note that seems worth fixing before merge: the new fail-closed engine validation still has one empty-config path that falls back to AES-CBC. The docs/style notes are smaller. All line numbers verified against HEAD `10c49b61`.

### Functional - worth fixing before merge

- **`superset/utils/encrypt.py:135`**

  The `or DEFAULT_ENCRYPTION_ENGINE_NAME` fallback means a present-but-empty `SQLALCHEMY_ENCRYPTED_FIELD_ENGINE` skips `resolve_encryption_engine()` and silently uses AES-CBC. That seems to undercut the fail-closed behavior documented in this same block: after a GCM migration, an empty env/config value could make new secrets get written as CBC while the operator thinks the app is still in GCM mode.

  WDYT - could we distinguish an absent key from a present empty value, and add a small test that `""` raises instead of falling back?

### Other suggestions

- **`docs/sip/authenticated-encryption-at-rest.md:111`**

  The SIP runbook still says backup -> run migrator -> set config -> restart, while `UPDATING.md` now adds a post-restart second pass to sweep up CBC writes during the cutover window. Since this file is the proposal/operators may cross-reference it from `UPDATING.md`, the two runbooks could leave readers unsure whether the second pass is required.

  Small suggestion: would it be worth syncing this runbook with the five-step `UPDATING.md` version?

- **`tests/integration_tests/cli_tests.py:371`**

  The new CLI tests in this block do not include `-> None` return annotations. Existing tests in this file predate that pattern, but the current repo guidance asks for type hints on new Python code/tests, and the unit tests added in this PR already follow it.

  Nit: could we add `-> None` to the new `test_re_encrypt_secrets_*` functions? Happy to keep as-is if this file intentionally stays untyped.

### Praise

- **`tests/unit_tests/utils/encrypt_test.py:256`**

  Nice coverage on the combined key-rotation plus engine-migration path. The test verifies that the rewritten value decrypts as AES-GCM under the current key, not just that the migrator issued an update.

**Comment by rusackas** [comment]
Thanks again @richardfogaca — all three caught and fixed in ad613d4.

The empty-value one was the real bug: `get(..., DEFAULT)` now only defaults when the key is *absent*, so a present-but-empty value routes through `resolve_encryption_engine` and fails closed like any other typo. Added a unit test that `""` raises. Synced the SIP runbook with the five-step `UPDATING.md` version (post-restart second pass + the quiet-window note) and pointed it at `UPDATING.md` as canonical. And added `-> None` to the new `test_re_encrypt_secrets_*` functions.

File: superset/cli/update.py [hunk 9]
```diff
@@ -110,17 +110,37 @@ def update_api_docs() -> None:
...
+    type=click.Choice(sorted(ENCRYPTION_ENGINES), case_sensitive=False),
+    help="Re-encrypt all app-encrypted fields with this encryption engine "
+    "(e.g. 'aes-gcm' for authenticated encryption). The SECRET_KEY is "
+    "unchanged. Take a metadata-DB backup first, then set "
+    "SQLALCHEMY_ENCRYPTED_FIELD_ENGINE to the same value and restart.",
+)
+def re_encrypt_secrets(
+    previous_secret_key: Optional[str] = None,
+    target_engine_name: Optional[str] = None,
+) -> None:
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/cli/update.py:127]: Good call, added a docstring.

File: superset/utils/encrypt.py [hunk 10]
```diff
@@ -101,10 +179,40 @@ def created_by_enc_field_factory(field: TypeDecorator) -> bool:
...
+    idempotency: a value already readable in the target form is left untouched,
+    so a run can be safely repeated or resumed.
+    """
+
+    def __init__(
+        self,
+        previous_secret_key: Optional[str] = None,
+        target_engine: Optional[type[Any]] = None,
+    ) -> None:
         from superset import db  # pylint: disable=import-outside-toplevel
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/utils/encrypt.py:224]: Added a docstring to the initializer.

File: superset/utils/encrypt.py [hunk 11]
```diff
@@ -19,17 +19,76 @@
...
+# authentication tag; AES-CBC does not, so a CBC "success" can be coincidental.
+# Classification logic (the migrator's idempotency fast path) must let an
+# authenticated decrypt win over an unauthenticated one, never the reverse.
+AUTHENTICATED_ENGINES: frozenset[type[EncryptionDecryptionBaseEngine]] = frozenset(
+    {AesGcmEngine}
+)
+
+
+def _is_authenticated_engine(engine: type[EncryptionDecryptionBaseEngine]) -> bool:
+    return engine in AUTHENTICATED_ENGINES
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/utils/encrypt.py:64]: Added a docstring here.

File: tests/integration_tests/cli_tests.py [hunk 12]
```diff
@@ -364,3 +366,94 @@ def test_re_encrypt_secrets_failure_exits_nonzero(app_context):
     # The failure path must be handled by the CLI, not leaked as an
     # uncaught exception.
     assert response.exception is None or isinstance(response.exception, SystemExit)
+
+
+def test_re_encrypt_secrets_engine_option_invokes_migrator(app_context) -> None:
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/integration_tests/cli_tests.py]: Annotated `app_context` as `AppContext`.

File: tests/integration_tests/cli_tests.py [hunk 13]
```diff
@@ -364,3 +366,94 @@ def test_re_encrypt_secrets_failure_exits_nonzero(app_context):
...
+            ["--engine", "aes-gcm"],
+        )
+
+    assert response.exit_code == 0
+    call_kwargs = migrator_mock.call_args.kwargs
+    assert call_kwargs.get("target_engine") is AesGcmEngine
+    assert call_kwargs.get("previous_secret_key") is None
+
+
+def test_re_encrypt_secrets_engine_option_case_insensitive(app_context) -> None:
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/integration_tests/cli_tests.py]: Annotated `app_context` as `AppContext`.

File: tests/integration_tests/cli_tests.py [hunk 14]
```diff
@@ -364,3 +366,94 @@ def test_re_encrypt_secrets_failure_exits_nonzero(app_context):
...
+        response = runner.invoke(
+            superset.cli.update.re_encrypt_secrets,
+            ["--engine", "AES-GCM"],
+        )
+
+    assert response.exit_code == 0
+    assert migrator_mock.call_args.kwargs.get("target_engine") is AesGcmEngine
+
+
+def test_re_encrypt_secrets_combined_key_rotation_and_engine(app_context) -> None:
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/integration_tests/cli_tests.py]: Annotated `app_context` as `AppContext`.

File: tests/integration_tests/cli_tests.py [hunk 15]
```diff
@@ -364,3 +366,94 @@ def test_re_encrypt_secrets_failure_exits_nonzero(app_context):
...
+            ["--previous_secret_key", "old-key", "--engine", "aes-gcm"],
+        )
+
+    assert response.exit_code == 0
+    call_kwargs = migrator_mock.call_args.kwargs
+    assert call_kwargs.get("target_engine") is AesGcmEngine
+    assert call_kwargs.get("previous_secret_key") == "old-key"
+
+
+def test_re_encrypt_secrets_engine_option_invalid_raises_usage(app_context) -> None:
```
[the remark this answers is not in the document]
Comment (rusackas) [reply tests/integration_tests/cli_tests.py]: Annotated `app_context` as `AppContext`.

File: superset/utils/encrypt.py [hunk 16]
```diff
@@ -19,17 +19,76 @@
...
+    operator who typo'd ``"aes_gcm"`` believe they had authenticated encryption,
+    and, after a GCM migration, write new secrets as CBC into a GCM database.
+
+    The offending value is deliberately kept out of the error message: it comes
+    from the app config (which also holds ``SECRET_KEY``), and static analysis
+    flags interpolating config-sourced values into logs/errors as potential
+    clear-text secret exposure. The set of valid engine names is enough to
+    diagnose a typo.
+    """
+    normalized = engine_name.strip().lower().replace("_", "-")
```
[the remark this answers is not in the document]
Comment (rusackas) [reply superset/utils/encrypt.py:93]: Good catch, `resolve_encryption_engine` now type-checks the value first so a non-string fails closed with the same `ValueError` instead of an `AttributeError`. Added a unit test.

**Review by richardfogaca — COMMENTED** [review]
_Posting on Richard's behalf — this is his PR reviewer agent. Forward any pushback to him and he'll loop me back in._

Left two notes below. The live-write overwrite case is the main thing I’d look at before merge; the test note is a smaller coverage suggestion. All line numbers verified against HEAD `e3f6631`.

### Functional — worth checking before merge

- **`superset/utils/encrypt.py:510`**

  The migrator decrypts the selected ciphertext earlier in the loop, then updates the row by primary key only. If a live worker changes the same secret after this command reads the row but before this `UPDATE` runs, this can write the migrated old plaintext back over the newer secret.

  That seems especially relevant because the runbook says a live instance may keep writing AES-CBC secrets during the first pass and that the second pass will sweep them up. The second pass catches rows written after the first migration, but it would not recover a newer same-row secret that was overwritten by the first pass.

  WDYT — would it be worth either documenting a real write freeze/maintenance window for secret writes, or adding an optimistic guard that includes the original ciphertext in the `WHERE` clause and treats changed rows as retry/failure?

### Other suggestions

- **`tests/unit_tests/utils/encrypt_test.py:51`**

  The engine-migration tests seed `SecretsMigrator` with `__new__`, and the CLI tests mock `SecretsMigrator` entirely. That leaves the real production wiring for `SecretsMigrator(previous_secret_key=None, target_engine=AesGcmEngine).run()` unexercised, including the constructor defaulting the previous key from `current_app.config["SECRET_KEY"]` and the transaction/discovery path.

  Small suggestion: could we add one integration-style test that runs the real migrator with `target_engine=AesGcmEngine` against an encrypted row? The row-level coverage is strong, but this would cover the operator path end to end.

### Praise

- **`tests/unit_tests/utils/encrypt_test.py:250`**

  The GCM-to-CBC rollback edge case is covered thoughtfully. Pinning that the authenticated probe wins over a spurious CBC decrypt is exactly the kind of invariant that would be easy to lose in a future cleanup.

**Comment by sha174n** [comment]
Reviewed the engine selection and the migrator in depth. The fail-closed resolver (an unknown engine raises at startup rather than degrading to CBC, with the config value kept out of the error) and the migrator's authenticated-wins classification (so a GCM value can't be mis-skipped on a CBC rollback) are both well done, and the run is correctly all-or-nothing (rolls back on any failure). Two non-blocking notes:

1. The whole re-encryption runs in a single `engine.begin()` transaction. That's the right safety tradeoff, but on an install with a very large encrypted table (e.g. many stored OAuth tokens) it's a long-held transaction; worth a caveat in the runbook or a future per-table batch option.

2. AES-GCM is randomized, so encrypted fields stop being queryable by value (unlike CBC). No functional break today since nothing queries encrypted columns by value, just worth keeping in mind for future code.

LGTM.

**Review by sha174n — APPROVED** [review]
Approving. Reviewed in depth: the fail-closed engine resolver, the migrator's authenticated-wins idempotency classification, and the all-or-nothing transaction (rolls back on any failure) all check out, and 15 unit tests pass. The two notes above (single-transaction footprint on very large tables, GCM non-queryability) are non-blocking.
